File: leet_code/linked_lists/num_connected_components.js
const numConnectComponents = (head, G) => {
  let cur = head,
      count = 0,
      found = false;

  while (cur) {
    if (G.includes(cur.val)) found = true;
    else if (found && !G.includes(cur.val)) {
      found = !found;
      count++;
    }
    cur = cur.next;
  }
  if (found) count++;
  return count;
};

File: leet_code/linked_lists/num_connected_components.rb
def num_connect_components(head, arr)
    cur = head
    count = 0
    found = false

    while cur
        if arr.include?(cur.val)
            found = true
        elsif found && !arr.include?(cur.val)
            found = !found
            count += 1
        end
        cur = cur.next
    end
    count += 1 if found
    count
end
